src/types.ts

```diff
@@ -0,0 +1,221 @@
+export interface ConfigOptions<TSchema extends Record<string, Config>, TNamespace extends string> {
+  /**
+   * Namespace of the configuration
+   *
+   * This namespace is used to consume the configuration from `window` and `localstorage`
+   */
+  namespace: string;
+
+  /**
+   * Schema of the configuration (used for runtime validation)
+   */
+  schema: TSchema;
+
+  /**
+   * Storage adapter
+   *
+   * @default window.localStorage
+   */
+  storage?: Storage;
+
+  /**
+   * Permit to overidde any config values in storage
+   *
+   * @default true
+   */
+  localOverride?: boolean;
+
+  /**
+   * Namespace for `useConfig()` return methods.
+   *
+   * Example:
+   * ```
+   *  // MyConfig.ts
+   *  export const { useConfig } = createConfig({
+   *    useConfigNamespace: "hello"
+   *   });
+   *
+   * // In a react component
+   * const {
+   *   getHelloConfig,
+   *   setHelloConfig,
+   *   getAllHelloConfig,
+   * } = useConfig();
+   * ```
+   */
+  useConfigNamespace?: TNamespace;
+}
+
+export type Config = StringConfig | NumberConfig | BooleanConfig | CustomConfig;
+
+export interface StringConfig {
+  type: "string";
+  enum?: string[];
+  default?: string | (() => string);
+  description?: string;
+}
+
+export interface StringEnumConfig extends StringConfig {
+  /**
+   * List of allowed values
+   */
+  enum: string[];
+}
```

Comment:
If `StringEnumConfig` extends `StringConfig` with non-optional `enum`, why `StringConfig` already has `enum?: string[];`.

Reply:
So you can do `field.type === "string" && field.enum`, otherwise this is painful to discriminate

need `as any` to access `enum:
![image](https://user-images.githubusercontent.com/1761469/114877076-3f59c080-9dff-11eb-8ed3-e2bb95b02fa0.png)

`enum` is not listed in the `field` properties
![image](https://user-images.githubusercontent.com/1761469/114877165-526c9080-9dff-11eb-9f4e-ca73509554d6.png)

Reply:
Could adding a separate field type `stringWithEnum` make it a bit clearer?